File: go/internal/sample/storage.go
package sample

import (
	"context"
	"database/sql"
	"errors"
	"sample-micro-service/pkg/service/database"
	"time"
)

type storedItem struct {
	id          string
	title       string
	description string
	status      string
	color       string
	iteration   int
	updatedAt   time.Time
}

type storedPreview struct {
	id        string
	data      string
	createdAt time.Time
}

func getItemsForUser(c context.Context, userId string, limit int) ([]storedItem, error) {
	rows, err := database.Db.QueryContext(c, "SELECT id, title, description, status, color, iteration, updated_at FROM go.items WHERE user_id = $1 ORDER BY id LIMIT $2", userId, limit)
	if err != nil {
		return make([]storedItem, 0), err
	}
	return fetchItems(rows)
}

func getItemsForUserFromId(c context.Context, userId string, fromId string, limit int) ([]storedItem, error) {
	rows, err := database.Db.QueryContext(c, "SELECT id, title, description, status, color, iteration, updated_at FROM go.items WHERE user_id = $1 AND id >= $2 ORDER BY id LIMIT $3", userId, fromId, limit)
	if err != nil {
		return make([]storedItem, 0), err
	}
	return fetchItems(rows)
}

func getItemForUser(c context.Context, userId string, itemId string) (*storedItem, error) {
	rows, err := database.Db.QueryContext(c, "SELECT id, title, description, status, color, iteration, updated_at FROM go.items WHERE user_id = $1 AND id = $2", userId, itemId)
	if err != nil {
		return nil, err
	}

	items, err := fetchItems(rows)
	if err != nil {
		return nil, err
	}

	if len(items) > 1 {
		return nil, errors.New("Too many items for " + itemId)
	}
	if len(items) < 1 {
		return nil, errors.New("Too few items for " + itemId)
	}
	return &items[0], nil
}

func getPreviewsForItem(c context.Context, itemId string) (result []storedPreview, err error) {
	result = make([]storedPreview, 0)

	rows, err := database.Db.QueryContext(c, "SELECT id, data, created_at FROM go.previews WHERE item_id = $1", itemId)
	if err != nil {
		return
	}

	for rows.Next() {
		var preview storedPreview
		err = rows.Scan(&preview.id, &preview.data, &preview.createdAt)
		if err != nil {
			return
		}
		result = append(result, preview)
	}
	return
}

func fetchItems(rows *sql.Rows) (result []storedItem, err error) {
	result = make([]storedItem, 0)

	for rows.Next() {
		var item storedItem
		err = rows.Scan(&item.id, &item.title, &item.description, &item.status, &item.color, &item.iteration, &item.updatedAt)
		if err != nil {
			return
		}
		result = append(result, item)
	}
	return
}

File: rust/src/sample/storage.rs
use chrono::NaiveDateTime;
use sqlx::Error;

use crate::shared::database::Database;

#[derive(sqlx::FromRow)]
pub struct StoredItem {
    pub id:  String,
    pub title: String ,
    pub description: String ,
    pub status: String ,
    pub color: String ,
    pub iteration: i64 ,
    pub updated_at: NaiveDateTime,
}

#[derive(sqlx::FromRow)]
pub struct StoredPreview {
    pub id:  String,
    pub data: String ,
    pub created_at: NaiveDateTime,
}

pub async fn get_stored_items(db: &Database, user_id: String, from: Option<String>, limit: i32) -> Result<Vec<StoredItem>, Error> {
    if from.is_none() {
        sqlx::query_as::<_, StoredItem>(
            "SELECT id, title, description, status, color, iteration, updated_at FROM rust.items WHERE user_id = $1 ORDER BY id LIMIT $2")
            .bind(user_id)
            .bind(limit)
            .fetch_all(&db.pool)
            .await
    } else {
        sqlx::query_as::<_, StoredItem>(
            "SELECT id, title, description, status, color, iteration, updated_at FROM rust.items WHERE user_id = $1 AND id >= $2 ORDER BY id LIMIT $3")
            .bind(user_id)
            .bind(from.unwrap())
            .bind(limit)
            .fetch_all(&db.pool)
            .await
    }
}

pub async fn get_stored_item(db: &Database, user_id: String, item_id: String) -> Result<Option<StoredItem>, Error> {
    sqlx::query_as::<_, StoredItem>(
        "SELECT id, title, description, status, color, iteration, updated_at FROM rust.items WHERE user_id = $1 AND id = $2")
        .bind(user_id)
        .bind(item_id)
        .fetch_optional(&db.pool)
        .await
}

pub async fn get_stored_previews(db: &Database, item_id: String) -> Result<Vec<StoredPreview>, Error> {
    sqlx::query_as::<_, StoredPreview>(
        "SELECT id, data, created_at FROM rust.previews WHERE item_id = $1")
        .bind(item_id)
        .fetch_all(&db.pool)
        .await
}
